// @chakra-ui/core/dist/es/Badge/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/** @jsx jsx */
import { jsx } from "@emotion/core";
import propTypes from "prop-types";
import useBadgeStyle from "./styles";
import Box from "../Box";
import { forwardRef } from "react";
var Badge = forwardRef(function (_ref, ref) {
  var variantColor = _ref.variantColor,
      variant = _ref.variant,
      props = _objectWithoutPropertiesLoose(_ref, ["variantColor", "variant"]);

  var badgeStyleProps = useBadgeStyle({
    color: variantColor,
    variant: variant
  });
  return jsx(Box, _extends({
    ref: ref,
    display: "inline-block",
    px: 1,
    textTransform: "uppercase",
    fontSize: "xs",
    borderRadius: "sm",
    fontWeight: "bold",
    whiteSpace: "nowrap",
    verticalAlign: "middle"
  }, badgeStyleProps, props));
});
process.env.NODE_ENV !== "production" ? Badge.propTypes = {
  variant: propTypes.oneOf(["solid", "subtle", "outline"])
} : void 0;
Badge.defaultProps = {
  variantColor: "gray",
  variant: "subtle"
};
export default Badge;
