// @chakra-ui/core/dist/NumberInput/Spinner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _Flex = _interopRequireDefault(require("../Flex"));

var _PseudoBox = _interopRequireDefault(require("../PseudoBox"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _ColorModeProvider = require("../ColorModeProvider");

var _styles = _interopRequireDefault(require("./styles"));

/** @jsx jsx */
var SpinButton = function SpinButton(_ref) {
  var isDisabled = _ref.isDisabled,
      onClick = _ref.onClick,
      onMouseDown = _ref.onMouseDown,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["isDisabled", "onClick", "onMouseDown"]);

  var _useColorMode = (0, _ColorModeProvider.useColorMode)(),
      colorMode = _useColorMode.colorMode;

  return (0, _core.jsx)(_PseudoBox["default"], (0, _extends2["default"])({
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    flex: "1",
    cursor: "pointer",
    transition: "all 0.3s",
    role: "button",
    tabindex: "-1",
    userSelect: "none",
    "aria-disabled": isDisabled,
    onClick: isDisabled ? null : onClick,
    onMouseDown: isDisabled ? null : onMouseDown
  }, (0, _styles["default"])({
    colorMode: colorMode
  }), props));
};

var Spinner = function Spinner(_ref2) {
  var incrementProps = _ref2.incrementProps,
      decrementProps = _ref2.decrementProps,
      iconSize = _ref2.iconSize,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["incrementProps", "decrementProps", "iconSize"]);
  return (0, _core.jsx)(_Flex["default"], (0, _extends2["default"])({
    flexDir: "column",
    "aria-hidden": true,
    w: "24px",
    m: "1px",
    pos: "absolute",
    right: "0px",
    h: "calc(100% - 2px)"
  }, rest), (0, _core.jsx)(SpinButton, incrementProps, (0, _core.jsx)(_Icon["default"], {
    name: "chevron-up",
    size: iconSize
  })), (0, _core.jsx)(SpinButton, decrementProps, (0, _core.jsx)(_Icon["default"], {
    name: "chevron-down",
    size: iconSize
  })));
};

var _default = Spinner;
exports["default"] = _default;
