// @chakra-ui/core/dist/es/Modal/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/** @jsx jsx */
import { jsx } from "@emotion/core";
import { ModalTransition, ModalOverlay, ModalContent } from "./components";

var Modal = function Modal(_ref) {
  var isOpen = _ref.isOpen,
      onClose = _ref.onClose,
      children = _ref.children,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? "md" : _ref$size,
      isCentered = _ref.isCentered,
      zIndex = _ref.zIndex,
      overlayBg = _ref.overlayBg,
      initialFocusRef = _ref.initialFocusRef,
      rest = _objectWithoutPropertiesLoose(_ref, ["isOpen", "onClose", "children", "size", "isCentered", "zIndex", "overlayBg", "initialFocusRef"]);

  var centeredStyle = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  };
  return jsx(ModalTransition, {
    isOpen: isOpen
  }, function (styles) {
    return jsx(ModalOverlay, {
      initialFocusRef: initialFocusRef,
      opacity: styles.opacity,
      bg: overlayBg,
      zIndex: zIndex,
      css: isCentered && centeredStyle,
      onDismiss: onClose
    }, jsx(ModalContent, _extends({
      maxWidth: size,
      rounded: "md",
      transform: "translate3d(0px, " + styles.y + "px, 0px)"
    }, !isCentered && {
      top: 40,
      mx: "auto"
    }, rest), children));
  });
};

export default Modal;
export * from "./components";
