// @chakra-ui/core/dist/Box/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = exports.systemProps = exports.truncate = void 0;

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _shouldForwardProp = _interopRequireDefault(require("@styled-system/should-forward-prop"));

var _styledSystem = require("styled-system");

var _config = _interopRequireDefault(require("./config"));

/** @jsx jsx */
var truncate = function truncate(props) {
  if (props.isTruncated) {
    return {
      overflow: "hidden",
      textOverflow: "ellipsis",
      whiteSpace: "nowrap"
    };
  }
};

exports.truncate = truncate;
var systemProps = (0, _styledSystem.compose)(_styledSystem.layout, _styledSystem.color, _styledSystem.space, _styledSystem.background, _styledSystem.border, _styledSystem.grid, _styledSystem.position, _styledSystem.shadow, _styledSystem.typography, _styledSystem.flexbox);
exports.systemProps = systemProps;
var Box = (0, _styled["default"])("div", {
  shouldForwardProp: _shouldForwardProp["default"]
})(truncate, systemProps, _config["default"]);
var _default = Box;
exports["default"] = _default;
