// @chakra-ui/core/dist/es/Button/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/** @jsx jsx */
import { jsx } from "@emotion/core";
import propTypes from "prop-types";
import { forwardRef } from "react";
import Icon from "../Icon";
import Spinner from "../Spinner";
import useButtonStyle from "./styles";
import PseudoBox from "../PseudoBox";
import Box from "../Box";

var ButtonIcon = function ButtonIcon(_ref) {
  var icon = _ref.icon,
      props = _objectWithoutPropertiesLoose(_ref, ["icon"]);

  if (typeof icon === "string") {
    return jsx(Icon, _extends({
      focusable: "false",
      name: icon,
      color: "currentColor"
    }, props));
  }

  return jsx(Box, _extends({
    as: icon,
    "data-custom-icon": true,
    focusable: "false",
    color: "currentColor"
  }, props));
};

var Button = forwardRef(function (_ref2, ref) {
  var isDisabled = _ref2.isDisabled,
      isLoading = _ref2.isLoading,
      isActive = _ref2.isActive,
      isFullWidth = _ref2.isFullWidth,
      children = _ref2.children,
      Comp = _ref2.as,
      variantColor = _ref2.variantColor,
      leftIcon = _ref2.leftIcon,
      rightIcon = _ref2.rightIcon,
      variant = _ref2.variant,
      loadingText = _ref2.loadingText,
      iconSpacing = _ref2.iconSpacing,
      type = _ref2.type,
      size = _ref2.size,
      colorMode = _ref2.colorMode,
      rest = _objectWithoutPropertiesLoose(_ref2, ["isDisabled", "isLoading", "isActive", "isFullWidth", "children", "as", "variantColor", "leftIcon", "rightIcon", "variant", "loadingText", "iconSpacing", "type", "size", "colorMode"]);

  var buttonStyleProps = useButtonStyle({
    color: variantColor,
    variant: variant,
    size: size,
    colorMode: colorMode
  });

  var _isDisabled = isDisabled || isLoading;

  return jsx(PseudoBox, _extends({
    disabled: _isDisabled,
    "aria-disabled": _isDisabled,
    ref: ref,
    as: Comp,
    type: type,
    borderRadius: "md",
    fontWeight: "semibold",
    width: isFullWidth ? "full" : undefined,
    "data-active": isActive ? "true" : undefined
  }, buttonStyleProps, rest), leftIcon && !isLoading && jsx(ButtonIcon, {
    ml: -1,
    mr: iconSpacing,
    icon: leftIcon
  }), isLoading && jsx(Spinner, {
    position: loadingText ? "relative" : "absolute",
    mr: loadingText ? iconSpacing : 0,
    color: "currentColor",
    size: "1em"
  }), isLoading ? loadingText || jsx(Box, {
    as: "span",
    opacity: "0"
  }, children) : children, rightIcon && !isLoading && jsx(ButtonIcon, {
    mr: -1,
    ml: iconSpacing,
    icon: rightIcon
  }));
});
Button.defaultProps = {
  variantColor: "gray",
  variant: "solid",
  size: "md",
  type: "button",
  iconSpacing: 2,
  as: "button"
};
process.env.NODE_ENV !== "production" ? Button.propTypes = {
  /**
   * The color of the button. Use the colors passed in `theme.colors`.
   */
  variantColor: propTypes.string,

  /**
   * The variant of the button style to use.
   */
  variant: propTypes.oneOf(["outline", "ghost", "unstyled", "link", "solid"]),

  /**
   * If `true`, the button will be disabled.
   */
  isDisabled: propTypes.bool,

  /**
   * If `true`, the button will show a spinner.
   */
  isLoading: propTypes.bool,

  /**
   * The label to show in the button when `isLoading` is true
   * If no text is passed, it only shows the spinner
   */
  loadingText: propTypes.string,

  /**
   * If `true`, the button will take up the full width of its container.
   */
  isFullWidth: propTypes.bool,

  /**
   * The html button type to use.
   */
  type: propTypes.oneOf(["button", "reset", "submit"]),

  /**
   * The size of the button. Use the sizes in `theme.sizes.button`
   */
  size: propTypes.oneOf(["xs", "sm", "md", "lg"]),

  /**
   * The content of the button.
   */
  children: propTypes.node.isRequired,

  /**
   * If added, the button will show an icon before the button's label.
   * Use the icon key in `theme.icons`
   */
  leftIcon: propTypes.oneOfType([propTypes.string, propTypes.func, propTypes.object]),

  /**
   * If added, the button will show an icon after the button's label.
   * Use the icon key in `theme.icons`
   */
  rightIcon: propTypes.oneOfType([propTypes.string, propTypes.func, propTypes.object]),

  /**
   * The space between the button icon and label.
   */
  iconSpacing: propTypes.oneOfType([propTypes.number, propTypes.string])
} : void 0;
export default Button;
