// @chakra-ui/core/dist/Image/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = exports.useHasImageLoaded = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _react = require("react");

var _Box = _interopRequireDefault(require("../Box"));

/** @jsx jsx */
var useHasImageLoaded = function useHasImageLoaded(_ref) {
  var src = _ref.src,
      onLoad = _ref.onLoad,
      onError = _ref.onError;

  var _useState = (0, _react.useState)(false),
      hasLoaded = _useState[0],
      setHasLoaded = _useState[1];

  (0, _react.useEffect)(function () {
    var image = new Image();
    image.src = src;

    var removeEventListeners = function removeEventListeners() {
      image.removeEventListener("load", loadListener);
      image.removeEventListener("error", errorListener);
    };

    var loadListener = function loadListener() {
      removeEventListeners();
      setHasLoaded(true);
      onLoad && onLoad();
    };

    var errorListener = function errorListener(err) {
      removeEventListeners();
      setHasLoaded(false);
      onError && onError(err);
    };

    image.addEventListener("load", loadListener);
    image.addEventListener("error", errorListener);
    return function () {
      if (hasLoaded) {
        return;
      }

      image.src = "";
    };
  }, [hasLoaded, src, onLoad, onError]);
  return hasLoaded;
};

exports.useHasImageLoaded = useHasImageLoaded;

var Img = function Img(_ref2) {
  var src = _ref2.src,
      onLoad = _ref2.onLoad,
      onError = _ref2.onError,
      fallbackSrc = _ref2.fallbackSrc,
      alt = _ref2.alt,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["src", "onLoad", "onError", "fallbackSrc", "alt"]);
  var hasLoaded = useHasImageLoaded({
    src: src,
    onLoad: onLoad,
    onError: onError
  });
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    as: "img",
    src: hasLoaded ? src : fallbackSrc
  }, props));
};

var _default = Img;
exports["default"] = _default;
