// @chakra-ui/core/dist/Badge/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = _interopRequireDefault(require("./styles"));

var _Box = _interopRequireDefault(require("../Box"));

var _react = require("react");

/** @jsx jsx */
var Badge = (0, _react.forwardRef)(function (_ref, ref) {
  var variantColor = _ref.variantColor,
      variant = _ref.variant,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["variantColor", "variant"]);
  var badgeStyleProps = (0, _styles["default"])({
    color: variantColor,
    variant: variant
  });
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    ref: ref,
    display: "inline-block",
    px: 1,
    textTransform: "uppercase",
    fontSize: "xs",
    borderRadius: "sm",
    fontWeight: "bold",
    whiteSpace: "nowrap",
    verticalAlign: "middle"
  }, badgeStyleProps, props));
});
process.env.NODE_ENV !== "production" ? Badge.propTypes = {
  variant: _propTypes["default"].oneOf(["solid", "subtle", "outline"])
} : void 0;
Badge.defaultProps = {
  variantColor: "gray",
  variant: "subtle"
};
var _default = Badge;
exports["default"] = _default;
