// @chakra-ui/core/dist/Menu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  useMenuContext: true,
  MenuButton: true,
  MenuDivider: true,
  MenuGroup: true,
  MenuList: true,
  MenuItem: true
};
exports.useMenuContext = useMenuContext;
exports.MenuItem = exports.MenuList = exports.MenuGroup = exports.MenuDivider = exports.MenuButton = exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _core = require("@emotion/core");

var _autoId = require("@reach/auto-id");

var _propTypes = require("prop-types");

var _react = require("react");

var _reactPopper = require("react-popper");

var _Box = _interopRequireDefault(require("../Box"));

var _PseudoBox = _interopRequireDefault(require("../PseudoBox"));

var _Text = _interopRequireDefault(require("../Text"));

var _ColorModeProvider = require("../ColorModeProvider");

var _usePrevious = _interopRequireDefault(require("../usePrevious"));

var _utils = require("../utils");

var _styles = require("./styles");

var _Divider = _interopRequireDefault(require("../Divider"));

var _MenuOption = require("./MenuOption");

Object.keys(_MenuOption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _MenuOption[key];
});

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var MenuContext = (0, _react.createContext)();

var Menu = function Menu(_ref) {
  var children = _ref.children,
      isOpen = _ref.isOpen,
      _ref$autoSelect = _ref.autoSelect,
      autoSelect = _ref$autoSelect === void 0 ? true : _ref$autoSelect,
      _ref$closeOnBlur = _ref.closeOnBlur,
      closeOnBlur = _ref$closeOnBlur === void 0 ? true : _ref$closeOnBlur,
      _ref$closeOnSelect = _ref.closeOnSelect,
      closeOnSelect = _ref$closeOnSelect === void 0 ? true : _ref$closeOnSelect;

  var _useColorMode = (0, _ColorModeProvider.useColorMode)(),
      colorMode = _useColorMode.colorMode;

  var _useState = (0, _react.useState)({
    isOpen: isOpen || false,
    index: -1
  }),
      state = _useState[0],
      setState = _useState[1];

  var menuId = "menu-" + (0, _autoId.useId)();
  var buttonId = "menubutton-" + (0, _autoId.useId)();
  var menuRef = (0, _react.useRef)(null);
  var buttonRef = (0, _react.useRef)(null);
  var focusableItems = (0, _react.useRef)(null);
  (0, _react.useEffect)(function () {
    var focusables = (0, _utils.getFocusables)(menuRef.current).filter(function (node) {
      return ["menuitem", "menuitemradio", "menuitemcheckbox"].includes(node.getAttribute("role"));
    });
    focusableItems.current = menuRef.current ? focusables : [];
    initTabIndex();
  }, []);

  var updateTabIndex = function updateTabIndex(index) {
    if (focusableItems.current.length > 0) {
      var nodeAtIndex = focusableItems.current[index];
      focusableItems.current.forEach(function (node) {
        if (node !== nodeAtIndex) {
          node.setAttribute("tabindex", -1);
        }
      });
      nodeAtIndex.setAttribute("tabindex", 0);
    }
  };

  var resetTabIndex = function resetTabIndex() {
    focusableItems.current.forEach(function (node) {
      return node.setAttribute("tabindex", -1);
    });
  };

  var initTabIndex = function initTabIndex() {
    focusableItems.current.forEach(function (node, index) {
      return index === 0 && node.setAttribute("tabindex", 0);
    });
  };

  var wasPreviouslyOpen = (0, _usePrevious["default"])(state.isOpen);
  (0, _react.useEffect)(function () {
    if (state.index !== -1) {
      focusableItems.current[state.index].focus();
      updateTabIndex(state.index);
    }

    if (state.index === -1 && !state.isOpen && wasPreviouslyOpen) {
      buttonRef.current && buttonRef.current.focus();
    }

    if (state.index === -1 && state.isOpen) {
      menuRef.current && menuRef.current.focus();
    }
  }, [state, wasPreviouslyOpen]);

  var focusOnFirstItem = function focusOnFirstItem() {
    setState({
      isOpen: true,
      index: 0
    });
  };

  var openMenu = function openMenu() {
    setState(_objectSpread({}, state, {
      isOpen: true
    }));
  };

  var focusAtIndex = function focusAtIndex(index) {
    setState(_objectSpread({}, state, {
      index: index
    }));
  };

  var focusOnLastItem = function focusOnLastItem() {
    setState({
      isOpen: true,
      index: focusableItems.current.length - 1
    });
  };

  var closeMenu = function closeMenu() {
    setState({
      isOpen: false,
      index: -1
    });
    resetTabIndex();
  };

  var context = {
    state: state,
    focusAtIndex: focusAtIndex,
    focusOnLastItem: focusOnLastItem,
    focusOnFirstItem: focusOnFirstItem,
    closeMenu: closeMenu,
    buttonRef: buttonRef,
    menuRef: menuRef,
    focusableItems: focusableItems,
    menuId: menuId,
    buttonId: buttonId,
    openMenu: openMenu,
    autoSelect: autoSelect,
    closeOnSelect: closeOnSelect,
    closeOnBlur: closeOnBlur,
    colorMode: colorMode
  };
  return (0, _core.jsx)(MenuContext.Provider, {
    value: context
  }, (0, _core.jsx)(_reactPopper.Manager, null, typeof children === "function" ? children({
    isOpen: state.isOpen,
    onClose: closeMenu
  }) : children));
};

function useMenuContext() {
  var context = (0, _react.useContext)(MenuContext);

  if (context === undefined) {
    throw new Error("useMenuContext must be used within a MenuContext Provider");
  }

  return context;
} //////////////////////////////////////////////////////////////////////////////////////////


var PseudoButton = (0, _react.forwardRef)(function (props, ref) {
  return (0, _core.jsx)(_PseudoBox["default"], (0, _extends2["default"])({
    ref: ref,
    as: "button"
  }, props));
});
var MenuButton = (0, _react.forwardRef)(function (_ref2, _ref4) {
  var _onClick = _ref2.onClick,
      _onKeyDown = _ref2.onKeyDown,
      _ref2$as = _ref2.as,
      Comp = _ref2$as === void 0 ? PseudoButton : _ref2$as,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["onClick", "onKeyDown", "as"]);

  var _useMenuContext = useMenuContext(),
      isOpen = _useMenuContext.state.isOpen,
      focusOnLastItem = _useMenuContext.focusOnLastItem,
      focusOnFirstItem = _useMenuContext.focusOnFirstItem,
      closeMenu = _useMenuContext.closeMenu,
      menuId = _useMenuContext.menuId,
      buttonId = _useMenuContext.buttonId,
      autoSelect = _useMenuContext.autoSelect,
      openMenu = _useMenuContext.openMenu,
      buttonRef = _useMenuContext.buttonRef;

  return (0, _core.jsx)(_reactPopper.Reference, null, function (_ref3) {
    var referenceRef = _ref3.ref;
    return (0, _core.jsx)(Comp, (0, _extends2["default"])({
      "aria-haspopup": "menu",
      "aria-expanded": isOpen,
      "aria-controls": menuId,
      id: buttonId,
      role: "button",
      ref: function ref(node) {
        return (0, _utils.mergeRefs)([buttonRef, referenceRef, _ref4], node);
      },
      onClick: function onClick(event) {
        if (isOpen) {
          closeMenu();
        } else {
          autoSelect ? focusOnFirstItem() : openMenu();
        }

        if (_onClick) {
          _onClick(event);
        }
      },
      onKeyDown: function onKeyDown(event) {
        if (event.key === "ArrowDown") {
          event.preventDefault();
          focusOnFirstItem();
        }

        if (event.key === "ArrowUp") {
          event.preventDefault();
          focusOnLastItem();
        }

        if (_onKeyDown) {
          _onKeyDown(event);
        }
      }
    }, rest));
  });
}); //////////////////////////////////////////////////////////////////////////////////////////

exports.MenuButton = MenuButton;

var MenuList = function MenuList(_ref5) {
  var onKeyDown = _ref5.onKeyDown,
      onBlur = _ref5.onBlur,
      placement = _ref5.placement,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref5, ["onKeyDown", "onBlur", "placement"]);

  var _useMenuContext2 = useMenuContext(),
      _useMenuContext2$stat = _useMenuContext2.state,
      index = _useMenuContext2$stat.index,
      isOpen = _useMenuContext2$stat.isOpen,
      focusAtIndex = _useMenuContext2.focusAtIndex,
      focusOnFirstItem = _useMenuContext2.focusOnFirstItem,
      focusOnLastItem = _useMenuContext2.focusOnLastItem,
      closeMenu = _useMenuContext2.closeMenu,
      focusableItems = _useMenuContext2.focusableItems,
      buttonRef = _useMenuContext2.buttonRef,
      menuId = _useMenuContext2.menuId,
      buttonId = _useMenuContext2.buttonId,
      menuRef = _useMenuContext2.menuRef,
      closeOnBlur = _useMenuContext2.closeOnBlur;

  var handleKeyDown = function handleKeyDown(event) {
    var count = focusableItems.current.length;
    var nextIndex;

    if (event.key === "ArrowDown") {
      event.preventDefault();
      nextIndex = (index + 1) % count;
      focusAtIndex(nextIndex);
    } else if (event.key === "ArrowUp") {
      event.preventDefault();
      nextIndex = (index - 1 + count) % count;
      focusAtIndex(nextIndex);
    } else if (event.key === "Home") {
      focusOnFirstItem();
    } else if (event.key === "End") {
      focusOnLastItem();
    } else if (event.key === "Tab") {
      event.preventDefault();
    } else if (event.key === "Escape") {
      closeMenu();
    } // Set focus based on first character


    if (/^[a-z0-9_-]$/i.test(event.key)) {
      event.stopPropagation();
      event.preventDefault();
      var foundNode = focusableItems.current.find(function (item) {
        return item.textContent.toLowerCase().startsWith(event.key);
      });

      if (foundNode) {
        nextIndex = focusableItems.current.indexOf(foundNode);
        focusAtIndex(nextIndex);
      }
    }

    onKeyDown && onKeyDown(event);
  }; // Close the menu on blur


  var handleBlur = function handleBlur(event) {
    if (closeOnBlur && isOpen && menuRef.current && buttonRef.current && !menuRef.current.contains(event.relatedTarget) && !buttonRef.current.contains(event.relatedTarget)) {
      closeMenu();
    }

    onBlur && onBlur(event);
  };

  var styleProps = (0, _styles.useMenuListStyle)();
  return (0, _core.jsx)(_reactPopper.Popper, {
    placement: placement
  }, function (_ref6) {
    var _ref7 = _ref6.ref,
        popperStyle = _ref6.style;
    return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
      minW: "3xs",
      rounded: "md",
      role: "menu",
      ref: function ref(node) {
        return (0, _utils.mergeRefs)([menuRef, _ref7], node);
      },
      id: menuId,
      py: 2,
      pos: "absolute",
      "aria-labelledby": buttonId,
      onKeyDown: handleKeyDown,
      onBlur: handleBlur,
      tabIndex: -1,
      zIndex: "1",
      hidden: !isOpen,
      css: popperStyle
    }, styleProps, props));
  });
}; //////////////////////////////////////////////////////////////////////////////////////////


exports.MenuList = MenuList;
var MenuItem = (0, _react.forwardRef)(function (_ref8, ref) {
  var isDisabled = _ref8.isDisabled,
      _onClick2 = _ref8.onClick,
      _onMouseLeave = _ref8.onMouseLeave,
      _onKeyDown2 = _ref8.onKeyDown,
      _onMouseMove = _ref8.onMouseMove,
      _ref8$role = _ref8.role,
      role = _ref8$role === void 0 ? "menuitem" : _ref8$role,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref8, ["isDisabled", "onClick", "onMouseLeave", "onKeyDown", "onMouseMove", "role"]);

  var _useMenuContext3 = useMenuContext(),
      focusableItems = _useMenuContext3.focusableItems,
      focusAtIndex = _useMenuContext3.focusAtIndex,
      closeOnSelect = _useMenuContext3.closeOnSelect,
      closeMenu = _useMenuContext3.closeMenu;

  var styleProps = (0, _styles.useMenuItemStyle)();
  return (0, _core.jsx)(_PseudoBox["default"], (0, _extends2["default"])({
    as: "button",
    ref: ref,
    display: "flex",
    textDecoration: "none",
    color: "inherit",
    minHeight: "32px",
    alignItems: "center",
    textAlign: "left",
    outline: "none",
    px: 4,
    role: role,
    tabIndex: -1,
    disabled: isDisabled,
    "aria-disabled": isDisabled,
    onClick: function onClick(event) {
      if (isDisabled) {
        event.stopPropagation();
        event.preventDefault();
        return;
      }

      _onClick2 && _onClick2(event);
      closeOnSelect && closeMenu();
    },
    onMouseMove: function onMouseMove(event) {
      if (isDisabled) {
        event.stopPropagation();
        event.preventDefault();
        return;
      }

      var nextIndex = focusableItems.current.indexOf(event.currentTarget);
      focusAtIndex(nextIndex);
      _onMouseMove && _onMouseMove(event);
    },
    onMouseLeave: function onMouseLeave(event) {
      focusAtIndex(-1);
      _onMouseLeave && _onMouseLeave(event);
    },
    onKeyDown: function onKeyDown(event) {
      if (isDisabled) return;

      if (event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        _onClick2 && _onClick2();
        closeOnSelect && closeMenu();
      }

      _onKeyDown2 && _onKeyDown2(event);
    }
  }, styleProps, props));
});
exports.MenuItem = MenuItem;
process.env.NODE_ENV !== "production" ? MenuItem.propTypes = {
  isDisabled: _propTypes.bool,
  onKeyDown: _propTypes.func,
  onClick: _propTypes.func,
  onMouseMove: _propTypes.func,
  role: _propTypes.string
} : void 0; //////////////////////////////////////////////////////////////////////////////////////////

var MenuDivider = (0, _react.forwardRef)(function (props, ref) {
  return (0, _core.jsx)(_Divider["default"], (0, _extends2["default"])({
    ref: ref,
    orientation: "horizontal"
  }, props));
}); //////////////////////////////////////////////////////////////////////////////////////////

exports.MenuDivider = MenuDivider;
var MenuGroup = (0, _react.forwardRef)(function (_ref9, ref) {
  var children = _ref9.children,
      title = _ref9.title,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref9, ["children", "title"]);
  return (0, _core.jsx)(_Box["default"], {
    ref: ref,
    role: "group"
  }, title && (0, _core.jsx)(_Text["default"], (0, _extends2["default"])({
    mx: 4,
    my: 2,
    fontWeight: "semibold",
    fontSize: "sm"
  }, rest), title), children);
}); //////////////////////////////////////////////////////////////////////////////////////////

exports.MenuGroup = MenuGroup;
var _default = Menu;
exports["default"] = _default;
