// @chakra-ui/core/dist/Popover/components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.PopoverBody = exports.PopoverFooter = exports.PopoverHeader = exports.PopoverCloseButton = exports.PopoverContent = exports.popperStyle = exports.PopoverTransition = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _react = require("react");

var _renderprops = require("react-spring/renderprops.cjs");

var _CloseButton = _interopRequireDefault(require("../CloseButton"));

var _Box = _interopRequireDefault(require("../Box"));

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2["default"])(["\n  > [data-arrow] {\n    width: 0;\n    height: 0;\n    border-style: solid;\n    position: absolute;\n    margin: ", ";\n  }\n\n  &[data-placement^=\"top\"] {\n    margin-bottom: ", ";\n    transform-origin: bottom center;\n  }\n\n  &[data-placement^=\"top\"] > [data-arrow] {\n    border-width: ", " ", " 0 ", ";\n    border-left-color: transparent;\n    border-right-color: transparent;\n    border-bottom-color: transparent;\n    bottom: -5px;\n    left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n    box-shadow: 3px 3px 7px rgba(0, 0, 0, 0.07);\n  }\n\n  &[data-placement^=\"bottom\"] {\n    margin-top: ", ";\n    transform-origin: top center;\n  }\n\n  &[data-placement^=\"bottom\"] > [data-arrow] {\n    border-width: 0 ", " ", " ", ";\n    border-left-color: transparent;\n    border-right-color: transparent;\n    border-top-color: transparent;\n    top: -5px;\n    left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n    box-shadow: -2px -2px 5px rgba(0, 0, 0, 0.06);\n  }\n\n  &[data-placement^=\"right\"] {\n    margin-left: 5px;\n    transform-origin: left center;\n  }\n\n  &[data-placement^=\"right\"] > [data-arrow] {\n    border-width: ", " ", " ", " 0;\n    border-left-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    left: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n    box-shadow: -3px 3px 7px rgba(0, 0, 0, 0.07);\n  }\n\n  &[data-placement^=\"left\"] {\n    margin-right: ", ";\n    transform-origin: right center;\n  }\n\n  &[data-placement^=\"left\"] > [data-arrow] {\n    border-width: ", " 0 ", " ", ";\n    border-top-color: transparent;\n    border-right-color: transparent;\n    border-bottom-color: transparent;\n    right: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n    box-shadow: 3px -3px 7px rgba(0, 0, 0, 0.07);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var arrowSize = "5px";

var PopoverTransition = function PopoverTransition(_ref) {
  var isOpen = _ref.isOpen,
      _ref$duration = _ref.duration,
      duration = _ref$duration === void 0 ? 200 : _ref$duration,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["isOpen", "duration", "children"]);
  return (0, _core.jsx)(_renderprops.Transition, (0, _extends2["default"])({
    items: isOpen,
    from: {
      opacity: 0.01,
      scale: 0.75
    },
    enter: {
      opacity: 1,
      scale: 1
    },
    leave: {
      opacity: 0,
      scale: 0.75
    },
    config: _objectSpread({
      duration: duration
    }, rest.config)
  }, rest), function (isOpen) {
    return isOpen && function (style) {
      return children(style);
    };
  });
};

exports.PopoverTransition = PopoverTransition;
var popperStyle = (0, _core.css)(_templateObject(), arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize);
exports.popperStyle = popperStyle;
var PopoverContent = (0, _react.forwardRef)(function (props, ref) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    width: "100%",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    rounded: "lg",
    boxShadow: "lg",
    maxWidth: "xs",
    ref: ref,
    css: popperStyle
  }, props));
});
exports.PopoverContent = PopoverContent;

var PopoverCloseButton = function PopoverCloseButton(_ref2) {
  var onClick = _ref2.onClick,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref2, ["onClick"]);
  return (0, _core.jsx)(_CloseButton["default"], (0, _extends2["default"])({
    size: "sm",
    onClick: onClick,
    position: "absolute",
    rounded: "md",
    top: "12px",
    right: "12px",
    p: "6px"
  }, props));
};

exports.PopoverCloseButton = PopoverCloseButton;

var PopoverHeader = function PopoverHeader(props) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    px: "12px",
    boxShadow: "bottom",
    py: "8px",
    as: "header"
  }, props));
};

exports.PopoverHeader = PopoverHeader;

var PopoverFooter = function PopoverFooter(props) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    px: "12px",
    boxShadow: "top",
    py: "8px",
    as: "footer"
  }, props));
};

exports.PopoverFooter = PopoverFooter;

var PopoverBody = function PopoverBody(_ref3) {
  var isScrollable = _ref3.isScrollable,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref3, ["isScrollable"]);
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    flex: "1",
    px: "12px",
    py: "8px"
  }, isScrollable && {
    overflow: "hidden auto"
  }, props));
};

exports.PopoverBody = PopoverBody;
