// @chakra-ui/core/dist/Popover/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _autoId = require("@reach/auto-id");

var _portal = _interopRequireDefault(require("@reach/portal"));

var _react = require("react");

var _reactPopper = require("react-popper");

var _Box = _interopRequireDefault(require("../Box"));

var _ColorModeProvider = require("../ColorModeProvider");

var _useDisclosure2 = _interopRequireDefault(require("../useDisclosure"));

var _utils = require("../utils");

var _reactFocusLock = _interopRequireDefault(require("react-focus-lock"));

var _components = require("./components");

Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _components[key];
});

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var Popover = function Popover(_ref) {
  var controlledIsOpen = _ref.isOpen,
      defaultIsOpen = _ref.defaultIsOpen,
      _ref$maxWidth = _ref.maxWidth,
      maxWidth = _ref$maxWidth === void 0 ? "xs" : _ref$maxWidth,
      trigger = _ref.trigger,
      gutter = _ref.gutter,
      placement = _ref.placement,
      children = _ref.children,
      showArrow = _ref.showArrow,
      showCloseButton = _ref.showCloseButton,
      _ref$usePortal = _ref.usePortal,
      usePortal = _ref$usePortal === void 0 ? true : _ref$usePortal,
      onOpenChange = _ref.onOpenChange,
      _ref$trapFocus = _ref.trapFocus,
      trapFocus = _ref$trapFocus === void 0 ? false : _ref$trapFocus,
      _ref$closeOnBlur = _ref.closeOnBlur,
      closeOnBlur = _ref$closeOnBlur === void 0 ? true : _ref$closeOnBlur,
      _ref$closeOnEsc = _ref.closeOnEsc,
      closeOnEsc = _ref$closeOnEsc === void 0 ? true : _ref$closeOnEsc,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["isOpen", "defaultIsOpen", "maxWidth", "trigger", "gutter", "placement", "children", "showArrow", "showCloseButton", "usePortal", "onOpenChange", "trapFocus", "closeOnBlur", "closeOnEsc"]);

  var _useDisclosure = (0, _useDisclosure2["default"])(defaultIsOpen),
      isOpen = _useDisclosure.isOpen,
      onClose = _useDisclosure.onClose,
      onToggle = _useDisclosure.onToggle;

  var triggerRef = (0, _react.useRef)();
  var popperRef = (0, _react.useRef)();
  (0, _react.useEffect)(function () {
    onOpenChange && onOpenChange(isOpen);
  }, [isOpen, onOpenChange]);

  var handleBlur = function handleBlur(event) {
    if (!trapFocus && isOpen && popperRef.current && triggerRef.current && !popperRef.current.contains(event.relatedTarget) && !triggerRef.current.contains(event.relatedTarget)) {
      closeOnBlur && onClose();
    }
  };

  var _useColorMode = (0, _ColorModeProvider.useColorMode)(),
      colorMode = _useColorMode.colorMode;

  var _bgColor = colorMode === "light" ? "white" : "gray.700";

  var bg = rest.bg || rest.background || rest.backgroundColor || _bgColor;
  var popoverId = "popper-" + (0, _autoId.useId)();
  var PopperWrapper = usePortal ? _portal["default"] : _react.Fragment;
  return (0, _core.jsx)(_reactPopper.Manager, null, (0, _core.jsx)(_reactPopper.Reference, null, function (_ref2) {
    var referenceRef = _ref2.ref;
    return (0, _react.cloneElement)(trigger, {
      "aria-haspopup": "true",
      "aria-controls": popoverId,
      ref: function ref(node) {
        triggerRef.current = node;
        (0, _utils.assignRef)(referenceRef, node);
      },
      onClick: function onClick(event) {
        onToggle();
        trigger.props.onClick && trigger.props.onClick(event);
      }
    });
  }), (0, _core.jsx)(PopperWrapper, null, (0, _core.jsx)(_reactPopper.Popper, {
    placement: placement
  }, function (_ref3) {
    var _ref4 = _ref3.ref,
        popperStyle = _ref3.style,
        placement = _ref3.placement,
        arrowProps = _ref3.arrowProps;
    return (0, _core.jsx)(_components.PopoverTransition, {
      duration: 100,
      isOpen: isOpen
    }, function (styles) {
      return (0, _core.jsx)(_reactFocusLock["default"] // eslint-disable-next-line jsx-a11y/no-autofocus
      , {
        autoFocus: false,
        returnFocus: true
      }, (0, _core.jsx)(_components.PopoverContent, (0, _extends2["default"])({
        ref: function ref(node) {
          popperRef.current = node;
          (0, _utils.assignRef)(_ref4, node);
        },
        bg: bg,
        maxWidth: maxWidth,
        "data-placement": placement,
        id: popoverId,
        "aria-hidden": isOpen
      }, rest, {
        tabIndex: "-1",
        onBlur: handleBlur,
        css: _objectSpread({}, popperStyle, {
          transform: popperStyle.transform + " scale(" + styles.scale + ")",
          opacity: styles.opacity
        }),
        onKeyDown: function onKeyDown(event) {
          event.stopPropagation();

          if (event.key === "Escape" && closeOnEsc) {
            onClose && onClose();
          }
        }
      }), showCloseButton && (0, _core.jsx)(_components.PopoverCloseButton, {
        onClick: onClose
      }), typeof children === "function" ? children({
        isOpen: isOpen,
        onClose: onClose
      }) : children, showArrow && (0, _core.jsx)(_Box["default"], {
        borderColor: bg,
        "data-arrow": "",
        ref: arrowProps.ref,
        css: arrowProps.style
      })));
    });
  })));
};

var _default = Popover;
exports["default"] = _default;
