// @chakra-ui/core/dist/AlertDialog/components.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.AlertDialogContent = exports.AlertDialogOverlay = exports.AlertDialogFooter = exports.AlertDialogBody = exports.AlertDialogHeader = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _core = require("@emotion/core");

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _alertDialog = require("@reach/alert-dialog");

var _react = require("react");

var _Box = _interopRequireWildcard(require("../Box"));

var _config = _interopRequireDefault(require("../Box/config"));

var _ColorModeProvider = require("../ColorModeProvider");

var _Modal = require("../Modal");

/** @jsx jsx */
var AlertDialogHeader = function AlertDialogHeader(props) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    py: 4,
    px: 6,
    as: _alertDialog.AlertDialogLabel
  }, props));
};

exports.AlertDialogHeader = AlertDialogHeader;

var AlertDialogBody = function AlertDialogBody(props) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    px: 6,
    py: 2,
    as: _alertDialog.AlertDialogDescription
  }, props));
};

exports.AlertDialogBody = AlertDialogBody;

var AlertDialogFooter = function AlertDialogFooter(props) {
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    ml: "auto",
    p: 6
  }, props));
};

exports.AlertDialogFooter = AlertDialogFooter;
var StyledOverlay = (0, _styled["default"])(_alertDialog.AlertDialogOverlay)(_Box.systemProps, _config["default"]);

var AlertDialogOverlay = function AlertDialogOverlay(_ref) {
  var _ref$zIndex = _ref.zIndex,
      zIndex = _ref$zIndex === void 0 ? 5 : _ref$zIndex,
      _ref$bg = _ref.bg,
      bg = _ref$bg === void 0 ? "rgba(16,22,26,.7)" : _ref$bg,
      props = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["zIndex", "bg"]);
  return (0, _core.jsx)(StyledOverlay, (0, _extends2["default"])({
    position: "fixed",
    zIndex: zIndex,
    bottom: "0",
    top: "0",
    left: "0",
    right: "0",
    overflowY: "auto",
    bg: bg
  }, props));
};

exports.AlertDialogOverlay = AlertDialogOverlay;
var AlertDialogContent = (0, _react.forwardRef)(function (props, ref) {
  var _useColorMode = (0, _ColorModeProvider.useColorMode)(),
      colorMode = _useColorMode.colorMode;

  var styleProps = (0, _Modal.modalContentStyle)({
    colorMode: colorMode
  });
  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    as: _alertDialog.AlertDialogContent,
    width: "100%",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    ref: ref
  }, styleProps, props));
});
exports.AlertDialogContent = AlertDialogContent;
