// @chakra-ui/core/dist/AlertDialog/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _components = require("../Modal/components");

var _components2 = require("./components");

Object.keys(_components2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _components2[key];
});

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var AlertDialog = function AlertDialog(_ref) {
  var isOpen = _ref.isOpen,
      onClose = _ref.onClose,
      children = _ref.children,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? "md" : _ref$size,
      overlayBg = _ref.overlayBg,
      zIndex = _ref.zIndex,
      isCentered = _ref.isCentered,
      leastDestructiveRef = _ref.leastDestructiveRef,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["isOpen", "onClose", "children", "size", "overlayBg", "zIndex", "isCentered", "leastDestructiveRef"]);
  return (0, _core.jsx)(_components.ModalTransition, {
    isOpen: isOpen
  }, function (styles) {
    return (0, _core.jsx)(_components2.AlertDialogOverlay, {
      bg: overlayBg,
      zIndex: zIndex,
      leastDestructiveRef: leastDestructiveRef,
      css: _objectSpread({
        opacity: styles.opacity
      }, isCentered && {
        display: "flex",
        alignItems: "center",
        justifyContent: "center"
      }),
      onDismiss: onClose
    }, (0, _core.jsx)(_components2.AlertDialogContent, (0, _extends2["default"])({
      maxWidth: size,
      rounded: "md",
      transform: "translate3d(0px, " + styles.y + "px, 0px)"
    }, !isCentered && {
      top: 40,
      mx: "auto"
    }, rest), children));
  });
};

var _default = AlertDialog;
exports["default"] = _default;
