// @chakra-ui/core/dist/es/NumberInput/Spinner.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/** @jsx jsx */
import { jsx } from "@emotion/core";
import Flex from "../Flex";
import PseudoBox from "../PseudoBox";
import Icon from "../Icon";
import { useColorMode } from "../ColorModeProvider";
import styleProps from "./styles";

var SpinButton = function SpinButton(_ref) {
  var isDisabled = _ref.isDisabled,
      onClick = _ref.onClick,
      onMouseDown = _ref.onMouseDown,
      props = _objectWithoutPropertiesLoose(_ref, ["isDisabled", "onClick", "onMouseDown"]);

  var _useColorMode = useColorMode(),
      colorMode = _useColorMode.colorMode;

  return jsx(PseudoBox, _extends({
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    flex: "1",
    cursor: "pointer",
    transition: "all 0.3s",
    role: "button",
    tabindex: "-1",
    userSelect: "none",
    "aria-disabled": isDisabled,
    onClick: isDisabled ? null : onClick,
    onMouseDown: isDisabled ? null : onMouseDown
  }, styleProps({
    colorMode: colorMode
  }), props));
};

var Spinner = function Spinner(_ref2) {
  var incrementProps = _ref2.incrementProps,
      decrementProps = _ref2.decrementProps,
      iconSize = _ref2.iconSize,
      rest = _objectWithoutPropertiesLoose(_ref2, ["incrementProps", "decrementProps", "iconSize"]);

  return jsx(Flex, _extends({
    flexDir: "column",
    "aria-hidden": true,
    w: "24px",
    m: "1px",
    pos: "absolute",
    right: "0px",
    h: "calc(100% - 2px)"
  }, rest), jsx(SpinButton, incrementProps, jsx(Icon, {
    name: "chevron-up",
    size: iconSize
  })), jsx(SpinButton, decrementProps, jsx(Icon, {
    name: "chevron-down",
    size: iconSize
  })));
};

export default Spinner;
