// @chakra-ui/core/dist/es/RadioGroup/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

/** @jsx jsx */
import { jsx } from "@emotion/core";
import { Children, cloneElement, useState, useRef, forwardRef } from "react";
import { useId } from "@reach/auto-id";
import Box from "../Box";
/* 
  TODO:
  - Calling focus() on the radiogroup should focus on the selected option or first enabled option
*/

var RadioGroup = forwardRef(function (_ref, ref) {
  var onChange = _ref.onChange,
      name = _ref.name,
      color = _ref.color,
      size = _ref.size,
      defaultValue = _ref.defaultValue,
      isInline = _ref.isInline,
      valueProp = _ref.value,
      _ref$spacing = _ref.spacing,
      spacing = _ref$spacing === void 0 ? 2 : _ref$spacing,
      children = _ref.children,
      rest = _objectWithoutPropertiesLoose(_ref, ["onChange", "name", "color", "size", "defaultValue", "isInline", "value", "spacing", "children"]);

  var _useRef = useRef(valueProp != null),
      isControlled = _useRef.current;

  var _useState = useState(defaultValue || null),
      value = _useState[0],
      setValue = _useState[1];

  var _value = isControlled ? valueProp : value;

  var _onChange = function _onChange(event) {
    var _val = event.target.value;
    !isControlled && setValue(_val);
    onChange && onChange(event, _val);
  }; // If no name is passed, we'll generate a random, unique name


  var fallbackName = "radio-" + useId();

  var _name = name || fallbackName;

  var clones = Children.map(children, function (child, index) {
    var isLastRadio = children.length === index + 1;
    var spacingProps = isInline ? {
      mr: spacing
    } : {
      mb: spacing
    };
    return jsx(Box, _extends({
      display: isInline ? "inline-block" : "block"
    }, !isLastRadio && spacingProps), cloneElement(child, {
      size: child.props.size || size,
      color: child.props.color || color,
      name: _name,
      onChange: _onChange,
      isChecked: child.props.value === _value
    }));
  });
  return jsx(Box, _extends({
    role: "radiogroup"
  }, rest), clones);
});
export default RadioGroup;
