// @chakra-ui/core/dist/es/Popover/components.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/taggedTemplateLiteralLoose";
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  > [data-arrow] {\n    width: 0;\n    height: 0;\n    border-style: solid;\n    position: absolute;\n    margin: ", ";\n  }\n\n  &[data-placement^=\"top\"] {\n    margin-bottom: ", ";\n    transform-origin: bottom center;\n  }\n\n  &[data-placement^=\"top\"] > [data-arrow] {\n    border-width: ", " ", " 0 ", ";\n    border-left-color: transparent;\n    border-right-color: transparent;\n    border-bottom-color: transparent;\n    bottom: -5px;\n    left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n    box-shadow: 3px 3px 7px rgba(0, 0, 0, 0.07);\n  }\n\n  &[data-placement^=\"bottom\"] {\n    margin-top: ", ";\n    transform-origin: top center;\n  }\n\n  &[data-placement^=\"bottom\"] > [data-arrow] {\n    border-width: 0 ", " ", " ", ";\n    border-left-color: transparent;\n    border-right-color: transparent;\n    border-top-color: transparent;\n    top: -5px;\n    left: calc(50% - 5px);\n    margin-top: 0;\n    margin-bottom: 0;\n    box-shadow: -2px -2px 5px rgba(0, 0, 0, 0.06);\n  }\n\n  &[data-placement^=\"right\"] {\n    margin-left: 5px;\n    transform-origin: left center;\n  }\n\n  &[data-placement^=\"right\"] > [data-arrow] {\n    border-width: ", " ", " ", " 0;\n    border-left-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    left: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n    box-shadow: -3px 3px 7px rgba(0, 0, 0, 0.07);\n  }\n\n  &[data-placement^=\"left\"] {\n    margin-right: ", ";\n    transform-origin: right center;\n  }\n\n  &[data-placement^=\"left\"] > [data-arrow] {\n    border-width: ", " 0 ", " ", ";\n    border-top-color: transparent;\n    border-right-color: transparent;\n    border-bottom-color: transparent;\n    right: -5px;\n    top: calc(50% - 5px);\n    margin-left: 0;\n    margin-right: 0;\n    box-shadow: 3px -3px 7px rgba(0, 0, 0, 0.07);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/** @jsx jsx */
import { css, jsx } from "@emotion/core";
import { forwardRef } from "react";
import { Transition } from "react-spring/renderprops.cjs";
import CloseButton from "../CloseButton";
import Box from "../Box";
var arrowSize = "5px";
export var PopoverTransition = function PopoverTransition(_ref) {
  var isOpen = _ref.isOpen,
      _ref$duration = _ref.duration,
      duration = _ref$duration === void 0 ? 200 : _ref$duration,
      children = _ref.children,
      rest = _objectWithoutPropertiesLoose(_ref, ["isOpen", "duration", "children"]);

  return jsx(Transition, _extends({
    items: isOpen,
    from: {
      opacity: 0.01,
      scale: 0.75
    },
    enter: {
      opacity: 1,
      scale: 1
    },
    leave: {
      opacity: 0,
      scale: 0.75
    },
    config: _objectSpread({
      duration: duration
    }, rest.config)
  }, rest), function (isOpen) {
    return isOpen && function (style) {
      return children(style);
    };
  });
};
export var popperStyle = css(_templateObject(), arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize, arrowSize);
export var PopoverContent = forwardRef(function (props, ref) {
  return jsx(Box, _extends({
    width: "100%",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    rounded: "lg",
    boxShadow: "lg",
    maxWidth: "xs",
    ref: ref,
    css: popperStyle
  }, props));
});
export var PopoverCloseButton = function PopoverCloseButton(_ref2) {
  var onClick = _ref2.onClick,
      props = _objectWithoutPropertiesLoose(_ref2, ["onClick"]);

  return jsx(CloseButton, _extends({
    size: "sm",
    onClick: onClick,
    position: "absolute",
    rounded: "md",
    top: "12px",
    right: "12px",
    p: "6px"
  }, props));
};
export var PopoverHeader = function PopoverHeader(props) {
  return jsx(Box, _extends({
    px: "12px",
    boxShadow: "bottom",
    py: "8px",
    as: "header"
  }, props));
};
export var PopoverFooter = function PopoverFooter(props) {
  return jsx(Box, _extends({
    px: "12px",
    boxShadow: "top",
    py: "8px",
    as: "footer"
  }, props));
};
export var PopoverBody = function PopoverBody(_ref3) {
  var isScrollable = _ref3.isScrollable,
      props = _objectWithoutPropertiesLoose(_ref3, ["isScrollable"]);

  return jsx(Box, _extends({
    flex: "1",
    px: "12px",
    py: "8px"
  }, isScrollable && {
    overflow: "hidden auto"
  }, props));
};
