// @chakra-ui/core/dist/RadioGroup/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _react = require("react");

var _autoId = require("@reach/auto-id");

var _Box = _interopRequireDefault(require("../Box"));

/** @jsx jsx */

/* 
  TODO:
  - Calling focus() on the radiogroup should focus on the selected option or first enabled option
*/
var RadioGroup = (0, _react.forwardRef)(function (_ref, ref) {
  var onChange = _ref.onChange,
      name = _ref.name,
      color = _ref.color,
      size = _ref.size,
      defaultValue = _ref.defaultValue,
      isInline = _ref.isInline,
      valueProp = _ref.value,
      _ref$spacing = _ref.spacing,
      spacing = _ref$spacing === void 0 ? 2 : _ref$spacing,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["onChange", "name", "color", "size", "defaultValue", "isInline", "value", "spacing", "children"]);

  var _useRef = (0, _react.useRef)(valueProp != null),
      isControlled = _useRef.current;

  var _useState = (0, _react.useState)(defaultValue || null),
      value = _useState[0],
      setValue = _useState[1];

  var _value = isControlled ? valueProp : value;

  var _onChange = function _onChange(event) {
    var _val = event.target.value;
    !isControlled && setValue(_val);
    onChange && onChange(event, _val);
  }; // If no name is passed, we'll generate a random, unique name


  var fallbackName = "radio-" + (0, _autoId.useId)();

  var _name = name || fallbackName;

  var clones = _react.Children.map(children, function (child, index) {
    var isLastRadio = children.length === index + 1;
    var spacingProps = isInline ? {
      mr: spacing
    } : {
      mb: spacing
    };
    return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
      display: isInline ? "inline-block" : "block"
    }, !isLastRadio && spacingProps), (0, _react.cloneElement)(child, {
      size: child.props.size || size,
      color: child.props.color || color,
      name: _name,
      onChange: _onChange,
      isChecked: child.props.value === _value
    }));
  });

  return (0, _core.jsx)(_Box["default"], (0, _extends2["default"])({
    role: "radiogroup"
  }, rest), clones);
});
var _default = RadioGroup;
exports["default"] = _default;
