// @chakra-ui/core/dist/Radio/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ControlBox = _interopRequireDefault(require("../ControlBox"));

var _ColorModeProvider = require("../ColorModeProvider");

var _VisuallyHidden = _interopRequireDefault(require("../VisuallyHidden"));

var _styles = _interopRequireDefault(require("../Checkbox/styles"));

var _Flex = _interopRequireDefault(require("../Flex"));

var _Box = _interopRequireDefault(require("../Box"));

/** @jsx jsx */
var Radio = (0, _react.forwardRef)(function (_ref, ref) {
  var id = _ref.id,
      name = _ref.name,
      value = _ref.value,
      ariaLabel = _ref["aria-label"],
      ariaLabelledBy = _ref["aria-labelledby"],
      color = _ref.color,
      defaultIsChecked = _ref.defaultIsChecked,
      isChecked = _ref.isChecked,
      isFullWidth = _ref.isFullWidth,
      size = _ref.size,
      isDisabled = _ref.isDisabled,
      isInvalid = _ref.isInvalid,
      onChange = _ref.onChange,
      onBlur = _ref.onBlur,
      onFocus = _ref.onFocus,
      children = _ref.children,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["id", "name", "value", "aria-label", "aria-labelledby", "color", "defaultIsChecked", "isChecked", "isFullWidth", "size", "isDisabled", "isInvalid", "onChange", "onBlur", "onFocus", "children"]);

  var _useColorMode = (0, _ColorModeProvider.useColorMode)(),
      colorMode = _useColorMode.colorMode;

  var styleProps = (0, _styles["default"])({
    color: color,
    size: size,
    colorMode: colorMode,
    type: "radio"
  });
  return (0, _core.jsx)(_Flex["default"], (0, _extends2["default"])({
    as: "label",
    display: "inline-flex",
    verticalAlign: "top",
    htmlFor: id,
    alignItems: "center",
    width: isFullWidth ? "full" : undefined,
    cursor: isDisabled ? "not-allowed" : "pointer"
  }, rest), (0, _core.jsx)(_VisuallyHidden["default"], {
    as: "input",
    type: "radio",
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    id: id,
    ref: ref,
    name: name,
    value: value,
    "aria-invalid": isInvalid,
    defaultChecked: defaultIsChecked,
    onChange: onChange,
    onBlur: onBlur,
    onFocus: onFocus,
    checked: isChecked,
    disabled: isDisabled
  }), (0, _core.jsx)(_ControlBox["default"], (0, _extends2["default"])({}, styleProps, {
    type: "radio",
    rounded: "full"
  }), (0, _core.jsx)(_Box["default"], {
    bg: "currentColor",
    as: "span",
    rounded: "full",
    size: "50%"
  })), children && (0, _core.jsx)(_Box["default"], {
    ml: 2,
    fontSize: size,
    userSelect: "none",
    opacity: isDisabled ? 0.32 : 1
  }, children));
});
Radio.defaultProps = {
  size: "md",
  color: "blue"
};
process.env.NODE_ENV !== "production" ? Radio.propTypes = {
  /**
   * The aria-label attribute associated with the radio element
   */
  "aria-label": _propTypes["default"].string,

  /**
   * If `true`, the radio will be disabled
   */
  isDisabled: _propTypes["default"].bool,

  /**
   * If `true`, the radio is marked as invalid.
   * Changes style of unchecked state.
   */
  isInvalid: _propTypes["default"].bool,

  /**
   * The color of the radio when it's checked.
   * This should be one of the color keys in the theme (e.g."green", "red")
   */
  color: _propTypes["default"].string,

  /**
   * If `true`, the radio will be initially checked.
   */
  defaultChecked: _propTypes["default"].bool,

  /**
   * If `true`, the radio will be checked.
   * You'll need to pass `onChange` to update it's value (since it's now controlled)
   */
  isChecked: _propTypes["default"].bool,

  /**
   * id assigned to input
   */
  id: _propTypes["default"].string,

  /**
   * The name of the input field in a radio
   * (Useful for form submission).
   */
  name: _propTypes["default"].string,

  /**
   * The value to be used in the radio button.
   * This is the value that will be returned on form submission.
   */
  value: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number])
} : void 0;
var _default = Radio;
exports["default"] = _default;
