// @chakra-ui/core/dist/Modal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {};
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _core = require("@emotion/core");

var _components = require("./components");

Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _components[key];
});

/** @jsx jsx */
var Modal = function Modal(_ref) {
  var isOpen = _ref.isOpen,
      onClose = _ref.onClose,
      children = _ref.children,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? "md" : _ref$size,
      isCentered = _ref.isCentered,
      zIndex = _ref.zIndex,
      overlayBg = _ref.overlayBg,
      initialFocusRef = _ref.initialFocusRef,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref, ["isOpen", "onClose", "children", "size", "isCentered", "zIndex", "overlayBg", "initialFocusRef"]);
  var centeredStyle = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  };
  return (0, _core.jsx)(_components.ModalTransition, {
    isOpen: isOpen
  }, function (styles) {
    return (0, _core.jsx)(_components.ModalOverlay, {
      initialFocusRef: initialFocusRef,
      opacity: styles.opacity,
      bg: overlayBg,
      zIndex: zIndex,
      css: isCentered && centeredStyle,
      onDismiss: onClose
    }, (0, _core.jsx)(_components.ModalContent, (0, _extends2["default"])({
      maxWidth: size,
      rounded: "md",
      transform: "translate3d(0px, " + styles.y + "px, 0px)"
    }, !isCentered && {
      top: 40,
      mx: "auto"
    }, rest), children));
  });
};

var _default = Modal;
exports["default"] = _default;
